File: backend/app/api/v1/routes/query.py
from fastapi import APIRouter

from app.schemas.query import QueryRequest, QueryResponse
from app.engine.reasoning_engine import ReasoningEngine


router = APIRouter()


@router.post(
    "/query",
    response_model=QueryResponse,
)
def query(request: QueryRequest):

    engine = ReasoningEngine()

    proof = engine.process(request.question)

    anchor = proof.chain_anchor or {}

    print("========== CHAIN ANCHOR ==========")
    print("CHAIN ANCHOR TYPE:", type(anchor))
    print("CHAIN ANCHOR VALUE:", anchor)
    print("VERDICT:", proof.verdict.verdict)
    print("ROOT HASH:", proof.root_hash)
    print("==================================")

    anchored = anchor.get("status") == "anchored"

    return QueryResponse(
        status="success",
        record_id=proof.record_id,
        question=proof.question,
        answer=proof.generator_answer,
        confidence_score=proof.verdict.confidence_score,
        confidence_level=proof.verdict.confidence_level.value,
        root_hash=proof.root_hash,
        anchored=anchored,
        tx_hash=anchor.get("tx_hash"),
        block_number=anchor.get("block_number"),
    )

File: frontend/src/api/query.ts
import { api } from "./client";

export interface QueryRequest {question: string;}

export interface QueryResponse {
  status: string;
  record_id: string;
  question: string;
  answer: string;
  confidence_score: number;
  confidence_level: string;
  root_hash: string;
  anchored: boolean;
  tx_hash?: string | null;
  block_number?: number | null;
}

export interface ProofRecord {
  record_id: string;
  question: string;
  created_at: number;

  evidence: Array<{
    chunk_id: string;
    text: string;
    source: string;
    similarity: number;
    bm25_score: number;
    fused_score: number;
    document_hash: string;
  }>;

  generator_answer: string;

  challenges: Array<{
    challenge_id: string;
    target_claim: string;
    objection: string;
    severity: number;
    resolved: boolean;
    resolution_note?: string | null;
  }>;

  fact_checks: Array<{
    claim: string;
    supported: boolean;
    supporting_evidence_ids: string[];
    contradicting_evidence_ids: string[];
    notes: string;
  }>;

  verdict: {
    verdict: string;
    confidence_score: number;
    confidence_level: string;
    rationale: string;
  };

  stages: Array<{
    stage: string;
    started_at: number;
    finished_at: number;
    payload: Record<string, unknown>;
    input_hash: string;
    output_hash: string;
  }>;

  root_hash: string;

  chain_anchor?: {
    status?: string;
    tx_hash?: string;
    transaction_hash?: string;
    block_number?: number | null;
  } | null;
}

export interface VerificationResponse {
  record_id: string;
  status:
    | "fully_verified"
    | "anchor_pending"
    | "valid_local_only"
    | "blockchain_mismatch"
    | "tampered";

  local_proof: {
    valid: boolean;
    stored_root_hash: string;
    computed_root_hash: string;
    tampered_stage: string | null;
    message: string;
  };

  blockchain: {
    anchored: boolean;
    verified: boolean;
    status: string;
    tx_hash: string | null;
    record: {
      proof_id: string;
      root_hash: string;
      timestamp: number;
      verifier_signature: string;
      submitter: string;
    } | null;
  };
}

export async function verifyQuestion(
  question: string,
): Promise<QueryResponse> {
  const { data } = await api.post<QueryResponse>(
    "/query",
    { question },
  );

  return data;
}

export async function getRecord(
  recordId: string,
): Promise<ProofRecord> {
  const { data } = await api.get<ProofRecord>(
    `/records/${recordId}`,
  );

  return data;
}

export async function verifyRecord(
  recordId: string,
): Promise<VerificationResponse> {
  const { data } = await api.get<VerificationResponse>(
    `/records/${recordId}/verify`,
  );

  return data;
}